### ISSUE TITLE
Set model optimizer as a parameter during tuning

### ISSUE THREAD

**Issue by adwaykanhere** [facebook/Ax#823]
Hey,

I'm trying to use Ax with Pytorch and I'm following your official tutorial. I will need to run my algorithm using different optimizers (Adam, SGD, RMSprop) while also tuning these function's learning rate and momentum. 
Something like this:

```
best_parameters, values, experiment, model = optimize(parameters=[
        {"name": "lr", "type": "range", "bounds": [1e-6, 0.4], "log_scale": True},

        {"name": "momentum", "type": "range", "bounds": [0.0, 1.0]},

        {"name": "epoch", "type": "range", "bounds": [15, 40]},

        {"name": "optimizer", "type": "choice", "values": [optim.SGD(lr,momentum), optim.Adam(lr,momentum)]}], 
evaluation_function=train_evaluate, objective_name='accuracy')
```

What would be the best way to go about doing this?

Thanks!

**Comment by bernardbeckerman** [comment]
Hi @adwaykanhere, thanks for posting! I'm assuming you're talking about [this tutorial](https://ax.dev/tutorials/tune_cnn.html), is that correct? If so, what problems are you running into during usage?

**Comment by adwaykanhere** [comment]
Thanks for the quick response. I'm trying to tune between two(for now) optimizers and their parameters like learning rate and momentum to find the best optimizer and parameter set to maximize accuracy. I'm writing my optimize function as described in the snippet above but while running the code I get an error. 

My question is - What "values" can I assign the "optimizer" such that it chooses among several optimizer functions while calling the momentum/learning rate (which are also parameters that need to be tuned)?

**Comment by Balandat** [comment]
The issue is 
```
"values": [optim.SGD(lr,momentum), optim.Adam(lr,momentum)]}]
```

The parameter types must be basic Ax types, and cannot be arbitrary objects such as torch optimizer instances. So what you can do instead is choose a STRING parameter type and allow choices `["SGD", "Adam"]` in the search space definition, and in your code construct an optimizer object based on the string value you receive for the "optimizer" parameter.

**Comment by Balandat** [comment]
```
def train_evaluate(params):
    if params["optimizer"] == "SGD":
        optimizer = optim.SGD(lr=params["lr"], momentum=params["momentum"])
    elif params["optimizer"] == "Adam":
        optimizer = optim.Adam(lr=params["lr"], momentum=params["momentum"])
    else:
        raise NotImplementedError
```

for the search space definition just do this 
```
{"name": "optimizer", "type": "choice", "values": ["SGD", "Adam"]}
```

**Comment by adwaykanhere** [comment]
Thank you very much! I'm closing this issue now.
